### PR TITLE
Fix on-demand initialization race conditions [marshal-ilgen.c]

### PR DESCRIPTION
**Pull request by jaykrell**
Extracted from https://github.com/mono/mono/pull/18150 which reviewer said was too big.

### CODE REVIEW

File: mono/metadata/marshal-ilgen.c [hunk 1]
```diff
@@ -4702,7 +4732,7 @@ emit_marshal_vtype_ilgen (EmitMarshalContext *m, int argnum, MonoType *t,
 			static MonoMethod *to_oadate;
 
 			if (!to_oadate)
-				to_oadate = get_method_nofail (date_time_class, "ToOADate", 0, 0);
+				mono_atomic_store_seq (&to_oadate, get_method_nofail (date_time_class, "ToOADate", 0, 0));
```
Comment (lambdageek) [inline mono/metadata/marshal-ilgen.c]: What am I missing? Why isn't this one using `MONO_STATIC_POINTER_INIT`?

File: mono/metadata/marshal-ilgen.c [hunk 2]
```diff
@@ -4808,7 +4838,7 @@ emit_marshal_vtype_ilgen (EmitMarshalContext *m, int argnum, MonoType *t,
 
 			if (!((t->attrs & PARAM_ATTRIBUTE_IN) && !(t->attrs & PARAM_ATTRIBUTE_OUT))) {
 				if (!from_oadate)
-					from_oadate = get_method_nofail (date_time_class, "FromOADate", 1, 0);
+					mono_atomic_store_seq (&from_oadate, get_method_nofail (date_time_class, "FromOADate", 1, 0));
```
Comment (lambdageek) [inline mono/metadata/marshal-ilgen.c]: Same with this one.  the scope of `from_oadata` could be narrowed to just the `if (!((t->attrs & PARAM_ATTRIBUTE_IN) && ...)` and then could be inited with `MONO_STATIC_POINTER_INIT`, couldn't it?

File: mono/metadata/marshal-ilgen.c (hunk 1 above)
Comment (jaykrell) [reply mono/metadata/marshal-ilgen.c]: In some cases they were used across multiple scopes. Let me try again here. Good catch thank you.

File: mono/metadata/marshal-ilgen.c [hunk 3]
```diff
@@ -4716,6 +4741,14 @@ emit_marshal_vtype_ilgen (EmitMarshalContext *m, int argnum, MonoType *t,
 				if (!t->byref)
 					m->csig->params [argnum - m->csig->hasthis] = double_type;
 
+			    MONO_STATIC_POINTER_INIT (MonoMethod, to_oadate)
+
+				    to_oadate = get_method_nofail (date_time_class, "ToOADate", 0, 0);
+
+			    MONO_STATIC_POINTER_INIT_END (MonoMethod, to_oadate)
+
+			    g_assert (to_oadate);
```
Comment (jaykrell) [inline mono/metadata/marshal-ilgen.c]: Looks like indent wrong.

File: mono/metadata/marshal-ilgen.c [hunk 4]
```diff
@@ -4801,14 +4834,16 @@ emit_marshal_vtype_ilgen (EmitMarshalContext *m, int argnum, MonoType *t,
...
 
 			if (!t->byref)
 				break;
 
 			if (!((t->attrs & PARAM_ATTRIBUTE_IN) && !(t->attrs & PARAM_ATTRIBUTE_OUT))) {
-				if (!from_oadate)
-					from_oadate = get_method_nofail (date_time_class, "FromOADate", 1, 0);
+
+				MONO_STATIC_POINTER_INIT (MonoMethod, from_oadate)
+					from_oadate =  get_method_nofail (date_time_class, "FromOADate", 1, 0);
```
Comment (jaykrell) [inline mono/metadata/marshal-ilgen.c]: Accidental space.

**Comment by jaykrell** [comment]
@monojenkins squash

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
@monojenkins build failed
